visidata/canvas.py

```diff
@@ -676,10 +676,10 @@ def render_sync(self):
 Canvas.addCommand('t', 'stoggle-cursor', 'source.toggle(list(rowsWithin(plotterCursorBox)))')
 Canvas.addCommand('u', 'unselect-cursor', 'source.unselect(list(rowsWithin(plotterCursorBox)))')
 Canvas.addCommand(ENTER, 'dive-cursor', 'vs=copy(source); vs.rows=list(rowsWithin(plotterCursorBox)); vd.push(vs)')
-Canvas.addCommand('d', 'delete-cursor', 'source.delete(list(rowsWithin(plotterCursorBox))); reload()')
+Canvas.addCommand('d', 'delete-cursor', 'toDelete=list(rowsWithin(plotterCursorBox)); source.copyRows(toDelete); source.deleteBy(lambda r,toDelete=toDelete: r in toDelete); reload()')
```

Comment:
This is workable for now, but this is O(N*M) and we can do O(N*log(M)) (where N is the number of rows and M is the number of rows to delete).  We ought to have a helper function that takes an iterable of rows and deletes them; it would create a dict of `rowid(r): row` (like `_selectedRows` is maintained as) from the iterable, and give that to the closure passed to deleteBy.  If you have a set of 2M data points and highlight 1M with the cursor and delete them from the canvas, the difference will be noticeable.

Also notice that `deleteSelected` is `@asyncthread`, and `deleteBy` is not.  So this would actually cause VisiData to hang for awhile in the above scenario.  So our new helper function should be async, and we could then call it with `rowsWithin(plotterCursorBox)`.  But then we need to figure out how to reload() the canvas after the delete thread finishes.  This is needed for Canvas.deleteCursor but not for Sheet, so for now we should skip the generic helper function and just refactor this execstr into its own canvas.deleteCursor() function and mark it asyncthread.